### PR TITLE
updatehub-sdk: doc: add a doc for the crate

### PR DESCRIPTION
**Pull request by Asakiz**
Signed-off-by: asakiz <asakizin@gmail.com>

### CODE REVIEW

File: updatehub-sdk/src/listener.rs [hunk 1]
```diff
@@ -17,11 +17,13 @@ const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
 
 type CallbackFn = dyn Fn(Handler) -> Pin<Box<dyn Future<Output = Result<()>>>>;
 
+/// Struct that store the callback(s) for the state(s).
 #[derive(Default)]
 pub struct StateChangeListener {
     callbacks: HashMap<State, Vec<Box<CallbackFn>>>,
 }
 
+/// The state(s) of the agent that can be handle.
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:26]:
```suggestion
/// The state of the agent that can be handled.
```

File: updatehub-sdk/src/listener.rs [hunk 2]
```diff
@@ -17,11 +17,13 @@ const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
 
 type CallbackFn = dyn Fn(Handler) -> Pin<Box<dyn Future<Output = Result<()>>>>;
 
+/// Struct that store the callback(s) for the state(s).
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:20]:
```suggestion
/// Struct that store the callbacks for the states.
```

File: updatehub-sdk/src/listener.rs [hunk 3]
```diff
@@ -49,6 +51,8 @@ impl FromStr for State {
     }
 }
 
+/// Handler used to communicate with
+/// UpdateHub to call the specified state callback(s).
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:55]:
```suggestion
/// Handler used to communicate with UpdateHub
/// to call commands on the state callbacks.
```

File: updatehub-sdk/src/listener.rs [hunk 4]
```diff
@@ -73,6 +77,17 @@ impl StateChangeListener {
         StateChangeListener::default()
     }
 
+    /// Function that register callback(s) for the state(s).
+    /// # Example
+    ///
+    /// ```text
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:83]: This should be a `no_run` block with the irrelevant parts hidden so the compiler can still check them.

File: updatehub-sdk/src/listener.rs [hunk 5]
```diff
@@ -81,6 +96,10 @@ impl StateChangeListener {
         self.callbacks.entry(state).or_insert_with(Vec::new).push(Box::new(move |d| Box::pin(f(d))))
     }
 
+    /// Start the agent to listen on the specified socket
+    /// the path is
+    /// `/usr/share/updatehub/state-change-callbacks.d/
+    /// 10-updatehub-sdk-statechange-trigger`.
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:102]:
```suggestion
    /// Start the agent to listen for messages on the socket at
    /// `/usr/share/updatehub/state-change-callbacks.d/
    /// 10-updatehub-sdk-statechange-trigger`.
```

**Review by Jonathas-Conceicao — COMMENTED** [review]
Got some minor suggestions

File: updatehub-sdk/src/listener.rs [hunk 6]
```diff
@@ -73,6 +77,19 @@ impl StateChangeListener {
         StateChangeListener::default()
     }
 
+    /// Function that register callback(s) for the state(s).
+    /// # Example
+    ///
+    /// ```no_run
+    /// let mut listener = listener::StateChangeListener::default();
```
Comment (otavio) [inline updatehub-sdk/src/listener.rs:84]: Does this compile?

Comment (Asakiz) [reply updatehub-sdk/src/listener.rs:84]: No, needs wrote the whole function.

**Review by otavio — APPROVED** [review]

**Review by otavio — CHANGES_REQUESTED** [review]

File: updatehub-sdk/src/listener.rs [hunk 7]
```diff
@@ -17,11 +17,13 @@ const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
 
 type CallbackFn = dyn Fn(Handler) -> Pin<Box<dyn Future<Output = Result<()>>>>;
 
+/// Struct that store the callbacks for the states.
 #[derive(Default)]
 pub struct StateChangeListener {
```
[the remark this answers is not in the document]
Comment (Asakiz) [reply updatehub-sdk/src/listener.rs:22]: I agree, but looking when we import the crate could be a little redundant.
e.g: `updatehub_sdk::listener::Listener::new()`.

[the remark this answers is not in the document]
Comment (otavio) [reply updatehub-sdk/src/listener.rs:22]: So just `StateChange`?

[the remark this answers is not in the document]
Comment (Asakiz) [reply updatehub-sdk/src/listener.rs:22]: It seems more appropriate.

**Review by otavio — DISMISSED** [review]

File: updatehub-sdk/src/listener.rs [hunk 8]
```diff
@@ -67,12 +71,29 @@ impl Handler {
...
+    /// # let mut listener = listener::StateChange::default();
+    /// listener.on_state(listener::State::Download, |mut handler| async move {
+    ///     println!("function called when starting the Download state");
+    ///
+    ///     // Cancels the current download.
+    ///     handler.cancel().await
+    /// });
+    /// # listener.listen().await
+    /// # }
+    /// ```
```
Comment (Jonathas-Conceicao) [inline updatehub-sdk/src/listener.rs:96]: This block does not produce the expected output, checkout the generated doc page.

**Review by Jonathas-Conceicao — CHANGES_REQUESTED** [review]
We are missing documentation for `Handler::cancel` e `Handler::proceed`.

**Review by Jonathas-Conceicao — APPROVED** [review]

**Review by otavio — APPROVED** [review]

**Pull request merged**